### PR TITLE
Sidebar Search App migration Review Only PR

### PR DESCRIPTION
**Pull request by yu-tian113**
This PR is only for review.
[https://zendesk.atlassian.net/browse/OFFAPPS-877](https://zendesk.atlassian.net/browse/OFFAPPS-877)

### CODE REVIEW

File: src/javascript/modules/search.js [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+  constructor (client, appData, config) {
+    this._client = client
+    this._appData = appData
+    this._config = config
+    this._apis = {
+      brands: `${this._client._origin}/api/v2/brands.json`,
+      users: `${this._client._origin}/api/v2/group_memberships/assignable.json?include=users`,
+      search: `${this._client._origin}/api/v2/search.json?per_page=${PER_PAGE}&query=`
+    }
+    this._states = {
```
Comment (ocke) [inline src/javascript/modules/search.js]: I'm not a big fan of this, because it's hard to follow what the values are, where they come from, the variable is not idempotent, but depends on the flow.

With promises it's much easier to understand the dataflow. Or when it comes to getting information, instead of setting the data in certain cases, have a function collect the right information when you need it. That way it's very clear where information is coming from.

```
state = { isSomething: false }

on('click', 'element', function(event) {
  state.isSomething = event.target.checked
)

function other() {
  if (state.isSomething) doThing()
}
```

I find this quite obscure and this doesn't scale. Instead I can write this with much less code:
```
function other() {
  if (document.querySelector('element').checked) doThing()
}
```

This is just an example, but the point is, instead of *setting* information when you have it, deal with it from a *getting* point of view, or even, pass the data from function to function. That way the code because much cleaner.

File: src/javascript/modules/search.js [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    return searchTerm
+  }
+
+  /**
+   * Format and render search results
+   * @param {Object} data response of the search request
+   */
+  _handleSearchResults (data) {
+    Object.assign(
+      this._states,
```
Comment (ocke) [inline src/javascript/modules/search.js]: Why not simply pass this object to the render function?

File: src/javascript/modules/search.js [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+      {
+        brands: await this._getBrands(),
+        suggestions: await this._getSearchSuggestions(),
+        assignees: await this._getAssignees()
+      }
+    )
+    // render application markup
+    await this._render('.loader', getSearchTemplate)
+    this._appContainer = document.querySelector('.search-app')
```
Comment (ocke) [inline src/javascript/modules/search.js]: Why assign these 3 lines to `this`? Why not simply use `document.querySelector('.search-app')` when you need it?

so on line 152 it would become `document.querySelector('.search-app .advanced-options-wrapper')`

File: src/javascript/modules/search.js [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+import getResultsTemplate from '../../templates/results'
+import getSearchTemplate from '../../templates/search'
+
+const PER_PAGE = 10
+const MAX_HEIGHT = 1000
+
+class Search {
+  constructor (client, appData, config) {
+    this._client = client
+    this._appData = appData
```
Comment (ocke) [inline src/javascript/modules/search.js]: What is find funny about this is that it gets passed from `main.js` line 7 `App(client, appData)` to `ticket_sidebar` line 9 `super(client, data, configs)` and then passed to `search.js` line 16 where we assign it to a instance variable (which in the case of this code can be seen as a global, because there is only 1 instance).

So we pass from function to function to function, then assign it to a global and then read from the global in the 1 spot where we use it (line 85)

Why not simply assign to a global from the beginning and then not worry about it?

File: src/javascript/modules/search.js [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+   * @return {Promise} Resolved to a suggestions array
+   */
+  async _getSearchSuggestions () {
+    const searchSuggestions = []
+    const customFieldIDs = this._appData.metadata.settings.custom_fields.match(/\d+/g)
+    const isRelatedTicketsConfigOn = this._appData.metadata.settings.related_tickets
+    const ticketSubject = (await this._client.get('ticket.subject'))['ticket.subject']
+    // custom field suggestions
+    if (customFieldIDs) {
+      await customFieldIDs.reduce((p, id) => {
```
Comment (ocke) [inline src/javascript/modules/search.js]: you can replace this loop 1 api call
`client.get(['ticket.customField:custom_field_12', 'ticket.customField:custom_field_34', 'ticket.customField:custom_field_56'])`

File: src/javascript/modules/search.js [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    else if (target.classList.contains('page-link') && target.dataset.url) this._doTheSearch(event, target.dataset.url)
+    else if (target.classList.contains('ticket-link')) this._handleResultLinkClick(event)
+  }
+
+  /**
+   * Keyword field keydown handler, search when press enter key
+   * @param {Event} event
+   */
+  _handleKeydown (event) {
+    event.which === 13 && this._doTheSearch(event)
```
Comment (ocke) [inline src/javascript/modules/search.js]: 🤔

File: src/javascript/modules/search.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+  }
+
+  /**
+   * Format and render error message
+   * @param {Object} data response of the search request
+   */
+  _handleSearchFail (data) {
+    const response = JSON.parse(data.responseText)
+    let message = ''
+    if (response.error) { message = I18n.t(`global.error.${response.error}`) } else if (response.description) { message = response.description } else { message = I18n.t('global.error.message') }
```
Comment (ocke) [inline src/javascript/modules/search.js]: the line of code that keeps on giving 😆

File: webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
+const path = require('path')
+const CleanWebpackPlugin = require('clean-webpack-plugin')
+const CopyWebpackPlugin = require('copy-webpack-plugin')
+const ExtractTextPlugin = require('extract-text-webpack-plugin')
+const TranslationsPlugin = require('./webpack/translations-plugin')
+// const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
+
+module.exports = (env) => {
+  // const getEnvConfig = (dev, prod) => {
```
Comment (ocke) [inline webpack.config.js]: we don't need this anymore?

File: src/templates/iframe.html [hunk 9]
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8">
+    <link href="styles.css" rel="stylesheet">
+  </head>
+
+  <body>
+    <section data-main>
+      <img class="loader" src="dot.gif"/>
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/templates/iframe.html]: Where can I find this?

File: src/templates/search.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
+import I18n from '../javascript/lib/i18n.js'
+import {templatingLoop as loop} from '../javascript/lib/helpers.js'
+var template = function (args) {
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/templates/search.js]: good one

File: src/templates/results.js [hunk 11]
```diff
@@ -0,0 +1,135 @@
+import I18n from '../javascript/lib/i18n.js'
+import {templatingLoop as loop} from '../javascript/lib/helpers.js'
+let getTicketMarkup = (o) => {
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/templates/results.js]: good one

File: src/javascript/locations/ticket_sidebar.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import Search from '../modules/search'
+// ticket sidebar specific configs
+const configs = {
+  // requiredProperties: ['ticket.id', 'ticket.subject']
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/javascript/locations/ticket_sidebar.js]: don't think so, deleted.

File: package.json [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    "file-loader": "^1.1.9",
+    "html-webpack-plugin": "^2.30.1",
+    "js-yaml": "^3.11.0",
+    "postcss-import": "^11.1.0",
+    "postcss-loader": "^2.1.5",
+    "postcss-preset-env": "^5.1.0",
+    "precss": "^3.1.2",
+    "url-loader": "^0.6.2",
+    "webpack": "^3.11.0",
+    "webpack-bundle-analyzer": "^2.12.0"
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply package.json]: only for dev, deleted

File: spec/i18n.spec.js [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        t('one')
+      }).toThrow()
+    })
+    it('loadTranslations return en translations', function () {
+      expect(I18n.loadTranslations('en')).toBe(mocktranslationObject)
+    })
+    it('loadTranslations fallback to en', function () {
+      expect(I18n.loadTranslations('FAIL')).toBe(mocktranslationObject)
+    })
+    it('loadTranslations fail, returnm empty object', function () {
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply spec/i18n.spec.js]: ~m~

File: src/javascript/modules/search.js [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+const PER_PAGE = 10
+const MAX_HEIGHT = 1000
+
+class Search {
+  constructor (client, appData, config) {
+    this._client = client
+    this._appData = appData
+    this._config = config
+    this._apis = {
+      brands: `${this._client._origin}/api/v2/brands.json`,
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/javascript/modules/search.js]: You are right, this is handled by ZAFClient. Deleted.

File: src/javascript/modules/search.js [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+      users: `${this._client._origin}/api/v2/group_memberships/assignable.json?include=users`,
+      search: `${this._client._origin}/api/v2/search.json?per_page=${PER_PAGE}&query=`
+    }
+    this._states = {
+      showAdvancedOptions: false,
+      hasMultiplebBrands: false,
+      results: [],
+      isLoading: false
+    }
+    this._initializePromise = this._init()
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/javascript/modules/search.js]: Yes, since all APIs are async, need a way to tell app is fully rendered.
Happy to discuss other solutions you guys done!

File: src/javascript/modules/search.js [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+import getSearchTemplate from '../../templates/search'
+
+const PER_PAGE = 10
+const MAX_HEIGHT = 1000
+
+class Search {
+  constructor (client, appData, config) {
+    this._client = client
+    this._appData = appData
+    this._config = config
```
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/javascript/modules/search.js]: This is not used at the moment until we have difference between locations for this app.

File: src/javascript/modules/search.js (hunk 1 above)
Comment (yu-tian113) [reply src/javascript/modules/search.js]: IMO, maintain a global states object is one way to separate the logic and UI, making the code more testable. It's similar idea to state management like redux. 
By doing this, we can easily switch to any dom render/manipulation library(vanilla, lodash, jquery, even react), also making sure logic is not bond to templates(change the checkbox to a button will need a rewrite to your function)
Also, you don't want to call dom operation everytime which is quite expensive.

File: src/javascript/modules/search.js (hunk 4 above)
Comment (yu-tian113) [reply src/javascript/modules/search.js]: There probably better ways to do this, I'm happy to discuss.
The main idea is in the future, apps will be used in multiple locations, we can make use of class extension to set different configurations, for example these two endpoints do the same thing on support and chat platform `comment.appendText` vs `chat.postToChatTextArea`
`client` and `data` are only passed to an extended class constructor(dynamically required in `main.js`), that's it, no more transpassing. `config` is where we differentiate locations. 
So what you've noticed is only code structure(class extension), not actual flow. 
(transpiler may end up convert it to multiple functions though)

File: webpack.config.js (hunk 8 above)
Comment (yu-tian113) [reply webpack.config.js]: Deleted

File: src/javascript/modules/search.js (hunk 5 above)
Comment (yu-tian113) [reply src/javascript/modules/search.js]: Nice Nice

File: src/javascript/modules/search.js (hunk 7 above)
Comment (yu-tian113) [reply src/javascript/modules/search.js]: this block is retained from old code, we can improve it later

File: src/templates/iframe.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (vaevictis) [reply src/templates/iframe.html]: https://github.com/zendesk/linked_ticket_app/blob/master/src/images/spinner.gif

File: src/javascript/modules/search.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (vaevictis) [reply src/javascript/modules/search.js]: Yeah, if we could avoid modifying the code for testing purposes, it would be great. How to achieve it in this case though? Not sure.

File: src/templates/results.js [hunk 18]
```diff
@@ -0,0 +1,135 @@
+import I18n from '../javascript/lib/i18n.js'
```
Comment (ocke) [inline src/templates/results.js]: Something new has crossed @vaevictis and I. The template mechanism that you use doesn't escape values. If someone creates a ticket with subject/description javascript code, that code will be inserted into the app, and therefore outsiders can run javascript on agents' behalf.

File: src/javascript/lib/helpers.js [hunk 19]
```diff
@@ -24,3 +24,17 @@ export async function loopingPaginatedRequest (client, url, entityName, max = Nu
   }
   return results
 }
+
+export function escapeSpecialChars (str) {
```
Comment (vaevictis) [inline src/javascript/lib/helpers.js]: What about adding a couple of tests for this function? Nothing flash, but just checking a few examples in would be great.

File: src/javascript/modules/search.js (hunk 6 above)
Comment (vaevictis) [reply src/javascript/modules/search.js]: rewrite like `if (event.which === 13) this.blahblah()`
